relayer/relayer/src/handlers/azero.rs

```diff
@@ -0,0 +1,267 @@
+
+impl AlephZeroEventsHandler {
+    pub async fn run(
+        config: Arc<Config>,
+        eth_signed_connection: Arc<SignedEthConnection>,
+        mut azero_events_receiver: mpsc::Receiver<AzeroMostEvents>,
+        circuit_breaker_sender: broadcast::Sender<CircuitBreakerEvent>,
+        mut circuit_breaker_receiver: broadcast::Receiver<CircuitBreakerEvent>,
+    ) -> Result<CircuitBreakerEvent, AlephZeroEventsHandlerError> {
+        loop {
+            debug!("Ping");
+
+            select! {
+                cb_event = circuit_breaker_receiver.recv() => {
+                    warn!("Exiting due to a circuit breaker event {cb_event:?}");
+                    return Ok(cb_event?);
+                },
+
+                Some(azero_events) = azero_events_receiver.recv() => {
+                    let AzeroMostEvents {
+                        events,
+                        events_ack_sender,
+                    } = azero_events;
+
+                    info!("Received a batch of {} events", events.len());
+
+                    let mut tasks = JoinSet::new();
+                    for event in events {
+                        // spawn each handler in separate task as it's time consuming
+                        tasks.spawn(AlephZeroEventHandler::handle_event(
+                            event,
+                            Arc::clone(&config),
+                            Arc::clone(&eth_signed_connection),
+                        ));
+                    }
+
+                    // wait for all concurrent handler tasks to finish
+                    info!("Awaiting all handler tasks to finish");
+
+                    while !tasks.is_empty() {
+                        select! {
+                            cb_event = circuit_breaker_receiver.recv() => {
+                                warn!("Exiting due to a circuit breaker event {cb_event:?}");
+                                return Ok(cb_event?);
+                            },
+
+                            Some (result) = tasks.join_next() => {
+                                if let Err (why) = result {
+                                    circuit_breaker_sender.send(CircuitBreakerEvent::AlephZeroEventHandlerFailure)?;
+                                    warn!("Event handler failed {why:?}, exiting");
+                                    return Ok(CircuitBreakerEvent::AlephZeroEventHandlerFailure);
+                                }
+                            }
+                        }
+                    }
+
+                    info!("Acknowledging events batch");
+                    // marks the batch as done and releases the listener
+                    events_ack_sender
+                        .send(())
+                        .map_err(|_| AlephZeroEventsHandlerError::EventsAckReceiverDropped)?;
+                    info!("All events acknowledged");
+                }
+            }
+        }
+    }
+}
+
+async fn wait_for_eth_tx_finality(
```

Comment:
As discussed, it would be better to add a getter (in the contract) for checking if this guardian needs to sign the request and using https://docs.rs/ethers/latest/ethers/middleware/contract/struct.FunctionCall.html#method.block (in a loop):
1. check in the latest block if guardian's vote is needed
- if no: (1) wait for the time usually needed for finalization (2) check if now it is true also for the most recent finalized block
- if the vote is needed: submit it, wait the time needed, check if it's still needed etc.

Reply:
Similar for submitting to Aleph, but there we can skip the waiting part and simply submit, but with first checking that submitting the tx is needed

Reply:
Yes, agreed with Krzysiek. The version with checking is infinitely more robust against virtually any problem.

Reply:
You added the check, but the main logic for checking if the change has finalized hasn't changed - please change it to the version in which check that changes happened is in form of call `needSigning` in the latest `Finalized` block. (wait_for_eth_tx_finality should not be needed)

Reply:
it's added